### PR TITLE
Fix invalidchar nick

### LINKED ISSUE #654: Mysteriously some users get bridged as "M[m]"
One poor user was `@kaylons:matrix.org` who never made it onto IRC because their bridgedclient couldn't connect. However, it was trying to bridge them as M which sounds...wrong.

### PR DESCRIPTION
**Pull request by Half-Shot**
Fixes #654. This checks for invalid characters much earlier during templating and throws more aggressively if the nickname looks wrong. This change comes with a nice set of unit tests, too.

### CODE REVIEW

File: lib/bridge/MatrixHandler.js [hunk 1]
```diff
@@ -904,9 +904,15 @@ MatrixHandler.prototype._onJoin = Promise.coroutine(function*(req, event, user)
...
-                        room.server.getNick(bridgedClient.userId, event.content.displayname),
-                        false);
+                    try {
+                        const newNick = room.server.getNick(
+                            bridgedClient.userId, event.content.displayname
+                        );
+                        bridgedClient.changeNick(newNick, false);
+                    }
+                    catch (e) {
+                        req.log.warn(`Didn't change nick on the IRC side:${e}`);
```
Comment (turt2live) [inline lib/bridge/MatrixHandler.js]: can has space after `:`?

File: lib/irc/IrcServer.js [hunk 2]
```diff
@@ -433,10 +434,16 @@ IrcServer.prototype.getAliasFromChannel = function(channel) {
...
 IrcServer.prototype.getNick = function(userId, displayName) {
-    var localpart = userId.substring(1).split(":")[0];
-    var display = displayName || localpart;
-    var template = this.config.ircClients.nickTemplate;
-    var nick = template.replace(/\$USERID/g, userId);
+    const illegalChars = BridgedClient.illegalCharactersRegex;
+    let localpart = userId.substring(1).split(":")[0];
+    localpart = localpart.replace(illegalChars, "");
+    displayName = displayName ? displayName.replace(illegalChars, "") : undefined;
+    let display = [displayName, localpart].find((n) => Boolean(n));
```
Comment (turt2live) [inline lib/irc/IrcServer.js]: This is a neat trick that I didn't think of. Can it be a constant though? :P

File: spec/unit/IrcServer.spec.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            );
+            expect(server.getNick("@💩ケ:foobar", "wiggle")).toBe("M-wiggle");
+        });
+        it("should throw if no characters could be used", function() {
+            const server = new IrcServer("irc.foobar",
+                extend(true, IrcServer.DEFAULT_CONFIG, {})
+            );
+            expect(() => {server.getNick("@💩ケ:foobar", "💩ケ")}).toThrow();
+        });
+    });
```
Comment (turt2live) [inline spec/unit/IrcServer.spec.js]: should also have a test for lack of display name (ie: null/undefined)

Comment (Half-Shot) [reply spec/unit/IrcServer.spec.js]: Isn't that covered by the first test?

Comment (Half-Shot) [reply spec/unit/IrcServer.spec.js]: Oh, you mean invalid userid only.

**Comment by Half-Shot** [comment]
*reassigns*
